### ISSUE TITLE
Joblib.externals.loky.process_executor.BrokenProcessPool: A task has failed to un-serialize.  Please ensure that the arguments of the arguments are all picklable.

### ISSUE THREAD

**Issue by sidnicholas** [freqtrade/freqtrade#2000]
## Step 2: Describe your environment

  * Python Version: 3.68
  * CCXT version: 1.18.819
  * Branch: Master | Develop
  * Last Commit ID: ab6c2da961d63ba2b68818ad79b7beffd78e1cc5
  
This issue has to deal with running hyperopt.  

I am running my bot on a Google Cloud VM.  I was able to get hyperopt to run on 1 CPU core although it eventually crashed.  After taking a look at freqtrade.io I read that the bot crashes with 1 CPU core.

I upgraded my Google Cloud VM from 1 to 2 cores.  I did not make any changes to the code. 
 Now when I run the code with 2 cores I get the following error message.
Joblib.externals.loky.process_executor.BrokenProcessPool: A task has failed to un-serialize.  Please ensure that the arguments of the arguments are all picklable.

I downgraded my CPU core to 1 and the code worked.  There's some issue that I am experiencing when I use a VM with two cores.

I've included screenshots here for review.  Thanks for any help you can offer.
https://drive.google.com/file/d/1wb9UvrhSv3-geTrZy2fJ6z94IWSWOeIK/view?usp=sharing
https://drive.google.com/file/d/1NjUw3J0GC26sdn6tKjYqJCCxJ76LDBLR/view?usp=sharing

Any help you can offer is much appreciated.

**Comment by hroff-1902** [comment]
Try
```
$ rm user_data/hyperopt_results.pickle 
$ rm user_data/hyperopt_tickerdata.pkl 
```

**Comment by xmatthias** [comment]
the difference between 1 and more cores is if joblib needs to pickle the data to parallelize it.

If you get this error, it's most likely related to something you do in your custom-hyperopt file, maybe an import which cannot be resolved or something similar.

**Comment by sidnicholas** [comment]
Thanks for your the responses.  Also, I would like to ask that you forgive me if I ask basic questions.  My programming chops are not quite there yet.

I forgot to mentioned that I removed hyperopt_results.pickle and hyperopt_tickerdata.pks.  I also removed the data files from the binance folder.  Those actions did not resolve the problem.

I will check to check my custom custom_hyperopt file to see if there is anything that appears unusual.  If I posted my file is that something that you would be willing to take a look at?  I will review myself before posting.

If I find a resolution I will post it here.

Thanks!

**Comment by hroff-1902** [comment]
yeah, by the way, how does it work on your systems (with 1 and 2 cores) for Default Hyperot?

**Comment by sidnicholas** [comment]
Good point.  I have only been using Custom Hyperopt.  I have not tried Default Hyperopt.  I will try Default Hyperopt.

**Comment by hroff-1902** [comment]
Another experiment which may help localize the problem: copy default_hyperopt.py to another file and change the name of the Hyperopt class in it (with no more changes -- to DefaultHyperOpts2, for example)...  That's because the default one is loaded in a diff. manner than the custom hyperopts... You can also try sample_hyperopt.py and test_hyperopt.py without modifications... If they run okay -- that means something is wrong with your custom hyperopts...

I suggest you to remove pickling files (those two mentioned above) every time you change the environment (1->2 cores and back) or change the hyperopt implementation...

**Comment by sidnicholas** [comment]
I was able to run DefaultHyperOpts on two cores and it worked fine.  I'll try the experiment from hroff-1902 previous post to see if I get the same result.  If everything runs okay, I'll rebuild my custom hyperopt.

**Comment by sidnicholas** [comment]
I renamed DefaultHyperOpt to DefaultHyperOpt2 I also renamed the .py file and everything worked.  I will rebuild the custom hyperopt file.  Question:  Is it possible my default strategy file could be contributing to the original problem?  Should I look at rebuilding the custom strategy file as well or would that have nothing to do with the original problem?  Thanks!

**Comment by xmatthias** [comment]
well you should not change DefaultStrategy ... (and hyperopt does not load another one), so the problem is (most likely) in the custom-hyperopt file.

You could try to use different `--spaces ` to help identify the problem, since that will limit what code is used

**Comment by sidnicholas** [comment]
I found several typos in my custom hyperopt file.
I did not have the 'return dataframe' for my populate_buy_trend function.
I also did not not have 'if conditions:' for the reduce method for both the buy and sell triggers.
Everything seems to be working now.

Thanks for your help!

**Comment by hroff-1902** [comment]
> I found several typos in my custom hyperopt file.
> I did not have the 'return dataframe' for my populate_buy_trend function.
> I also did not not have 'if conditions:' for the reduce method for both the buy and sell triggers.
> Everything seems to be working now.
> 
> Thanks for your help!

@sidnicholas Fine, you are welcome!

@xmatthias Could/should we add here a kind of protection against custom code not returning data (checking for the datatype of returned data, if it's not None, for example)?

**Comment by xmatthias** [comment]
i think we should do "some" sort of validation of a strategy (not necessarily for hyperopt (but it may be applicable to that too) - since i still see that as something you should only do once you know how to build a strategy in and out ...

**Comment by sidnicholas** [comment]
I used a online course to build my strategy and hyperopt file using freqtrade.  The instruction was really good but it move a little fast in some places.  That's how I made my mistakes. 

Now that I'm a little more familiar with the code and I've reviewed the docs on freqtrade.io I'm a bit more familiar with the code.  

It was totally user error on my part although I think adding some kind adding strategy validation would not diminish the software.  I think it would improve it.  Thanks again.
